# Index Explained

***If your system doesn't have pandas installed, use ! before your script, it can run the cell like a command window ***

In [ ]:
! pip install pandas 

**Our exercise dataset can be downloaded directly with pandas help from uci ml data repository <br> **` Dataset Name: Adult `

In [ ]:
import pandas as pd

adult  = pd.read_csv("https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data",
                  names = ['age', 'workclass','fnlwgt', 'education', 'education_num','marital_status', 
                         'occupation','relationship', 'race', 'sex', 'capital_gain','capital_loss',
                         'hours_per_week', 'native_country','label'], index_col = False)

print("Shape of data{}".format(adult.shape))

adult.head()

Shape of data(32561, 15)


age          workclass  fnlwgt   education  education_num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital_status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital_gain  capital_loss  hours_per_week  native_country   label  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40   United-States   <=50K  
4             0             0              40            Cuba   <=50K

* Dataset has 32561 rows and 15 features, the left most series 0,1 2,3 ... is index.<br>

**Looking at little more information **

In [ ]:
adult.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
age               32561 non-null int64
workclass         32561 non-null object
fnlwgt            32561 non-null int64
education         32561 non-null object
education_num     32561 non-null int64
marital_status    32561 non-null object
occupation        32561 non-null object
relationship      32561 non-null object
race              32561 non-null object
sex               32561 non-null object
capital_gain      32561 non-null int64
capital_loss      32561 non-null int64
hours_per_week    32561 non-null int64
native_country    32561 non-null object
label             32561 non-null object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


* Adult has rangeindex 32561 entries, an integer series from 0 to 32560 

## Selection Methods
1. df.loc are for labels/ names
2. df.iloc are for position numbers

> e.g. Lets assume Ram, Sonu & Tony are standing at positions 1, 2 & 3 respectively. If you want to call Ram you have two options, either you call him by his name or his position. So, if you call Ram by his name "Ram", you will use df.loc and if we will call him by his position "1" we will use df.iloc. 



**Before we understand loc & iloc more, lets take a sample from our data for further analysis. We are taking a sample of 10000 observations, using pandas df.sample method **

In [ ]:
df = adult.sample(10000, random_state = 100).sort_index(axis = 0)
df.head()

age   workclass  fnlwgt   education  education_num  \
3    53     Private  234721        11th              7   
4    28     Private  338409   Bachelors             13   
6    49     Private  160187         9th              5   
8    31     Private   45781     Masters             14   
11   30   State-gov  141297   Bachelors             13   

            marital_status          occupation    relationship  \
3       Married-civ-spouse   Handlers-cleaners         Husband   
4       Married-civ-spouse      Prof-specialty            Wife   
6    Married-spouse-absent       Other-service   Not-in-family   
8            Never-married      Prof-specialty   Not-in-family   
11      Married-civ-spouse      Prof-specialty         Husband   

                   race      sex  capital_gain  capital_loss  hours_per_week  \
3                 Black     Male             0             0              40   
4                 Black   Female             0             0              40   
6                 Black   Female             0             0              16   
8                 White   Female         14084             0              50   
11   Asian-Pac-Islander     Male             0             0              40   

    native_country   label  
3    United-States   <=50K  
4             Cuba   <=50K  
6          Jamaica   <=50K  
8    United-States    >50K  
11           India    >50K

**Nice, now we have a dataset name df, and the left most series are 3,4,6,8, 11...Strange! ** <br>
* It is because the rows will carry their original( old address ) index or index names from the adult dataset

**Naming our index will help us a little initially, which we got from adult and look at the rows and column indices **

In [ ]:
df.index.name = 'index_adult'

In [ ]:
df.index

Int64Index([    3,     4,     6,     8,    11,    12,    16,    23,    26,
               29,
            ...
            32525, 32526, 32529, 32536, 32541, 32542, 32543, 32548, 32550,
            32556],
           dtype='int64', name='index_adult', length=10000)

In [ ]:
df.columns

Index(['age', 'workclass', 'fnlwgt', 'education', 'education_num',
       'marital_status', 'occupation', 'relationship', 'race', 'sex',
       'capital_gain', 'capital_loss', 'hours_per_week', 'native_country',
       'label'],
      dtype='object')

* Both rows and columns have indexes, and the name of index is 'index_adult' <br>

**Lets discuss couple of examples on loc & iloc methods **

In [ ]:
# Example 1 : select third row and age value
df.loc[2,['age']] # Try replace 3,4 instead of 2

KeyError: 'the label [2] is not in the [index]'

In [ ]:
#Example2 : select third row and age value
df.iloc[2,0]

49

* In the first example of .loc, it gave us error
 * because we have used .loc method and df has no row who has a name '2', row index looks like a number to us, but they are name/label to .loc method
 * try replacing 2 with 3 or 4, it will work because there are names '3','4' <br>
* In the second example we are trying same with .iloc, its a position based method 
 * "age" is first column so we will use its position which is 0
 * there will be a position 0,1,2,3 till the last row of df, so 2 will be the third row

**For our further analysis, lets Keep few interesting variables only **

In [ ]:
cols = ['age',  'education', 'marital_status', 'occupation', 'race', 'sex', 'native_country','label'] 
df = df[cols]
df.head()

age   education          marital_status          occupation  \
index_adult                                                                
3             53        11th      Married-civ-spouse   Handlers-cleaners   
4             28   Bachelors      Married-civ-spouse      Prof-specialty   
6             49         9th   Married-spouse-absent       Other-service   
8             31     Masters           Never-married      Prof-specialty   
11            30   Bachelors      Married-civ-spouse      Prof-specialty   

                            race      sex  native_country   label  
index_adult                                                        
3                          Black     Male   United-States   <=50K  
4                          Black   Female            Cuba   <=50K  
6                          Black   Female         Jamaica   <=50K  
8                          White   Female   United-States    >50K  
11            Asian-Pac-Islander     Male           India    >50K

**Some of the times, it's difficult to work with random numbers in index, at that time, resetting this index will make it a column and recreate an another default index ** <br>

`reset_index() will recreate index column every time we run it on same data ` <br>
`drop = True paramater won't create that as column in the dataframe, look at the difference between following two dataset`


In [ ]:
df.reset_index().head()

index_adult  age   education          marital_status          occupation  \
0            3   53        11th      Married-civ-spouse   Handlers-cleaners   
1            4   28   Bachelors      Married-civ-spouse      Prof-specialty   
2            6   49         9th   Married-spouse-absent       Other-service   
3            8   31     Masters           Never-married      Prof-specialty   
4           11   30   Bachelors      Married-civ-spouse      Prof-specialty   

                  race      sex  native_country   label  
0                Black     Male   United-States   <=50K  
1                Black   Female            Cuba   <=50K  
2                Black   Female         Jamaica   <=50K  
3                White   Female   United-States    >50K  
4   Asian-Pac-Islander     Male           India    >50K

In [ ]:
df.reset_index(drop = True).head()

age   education          marital_status          occupation  \
0   53        11th      Married-civ-spouse   Handlers-cleaners   
1   28   Bachelors      Married-civ-spouse      Prof-specialty   
2   49         9th   Married-spouse-absent       Other-service   
3   31     Masters           Never-married      Prof-specialty   
4   30   Bachelors      Married-civ-spouse      Prof-specialty   

                  race      sex  native_country   label  
0                Black     Male   United-States   <=50K  
1                Black   Female            Cuba   <=50K  
2                Black   Female         Jamaica   <=50K  
3                White   Female   United-States    >50K  
4   Asian-Pac-Islander     Male           India    >50K

**inplace = True save us from assigning it to data again **<br>

`we won't use drop = True, now df should have its last index as a column in it `

In [ ]:
df.reset_index(inplace = True)
df.head()

index_adult  age   education          marital_status          occupation  \
0            3   53        11th      Married-civ-spouse   Handlers-cleaners   
1            4   28   Bachelors      Married-civ-spouse      Prof-specialty   
2            6   49         9th   Married-spouse-absent       Other-service   
3            8   31     Masters           Never-married      Prof-specialty   
4           11   30   Bachelors      Married-civ-spouse      Prof-specialty   

                  race      sex  native_country   label  
0                Black     Male   United-States   <=50K  
1                Black   Female            Cuba   <=50K  
2                Black   Female         Jamaica   <=50K  
3                White   Female   United-States    >50K  
4   Asian-Pac-Islander     Male           India    >50K

**Now df has an another column index_adult, because of reset **

## Filtering 
**Filter on India **

In [ ]:
ind = df[df.native_country ==' India']
ind.head()

index_adult  age   education          marital_status        occupation  \
4              11   30   Bachelors      Married-civ-spouse    Prof-specialty   
312          1029   48     Masters   Married-spouse-absent             Sales   
637          2130   28        11th               Separated      Adm-clerical   
902          2989   44     Masters      Married-civ-spouse   Exec-managerial   
1365         4480   30   Bachelors      Married-civ-spouse             Sales   

                     race      sex native_country   label  
4      Asian-Pac-Islander     Male          India    >50K  
312    Asian-Pac-Islander     Male          India   <=50K  
637    Asian-Pac-Islander   Female          India   <=50K  
902    Asian-Pac-Islander     Male          India   <=50K  
1365   Asian-Pac-Islander     Male          India   <=50K

 * After resetting our index, and applying a filter for India, we can see index hold itself from train, just like sampling, now the row index(4,312 637..) are from train and index_adult is the indices of these rows in df 

**Lets look at the samples with more than 50k income across the gender **

In [ ]:
cross = pd.crosstab(ind.sex, ind.label)
cross

label     <=50K   >50K
sex                   
 Female       4      0
 Male        16     17

In [ ]:
cross.index

Index([' Female', ' Male'], dtype='object', name='sex')

 * Even this dataframe has an index, hard to recognise by looking at the dataframe, and individual items can be accessed like 

In [ ]:
cross.loc[' Male']

label
 <=50K    16
 >50K     17
Name:  Male, dtype: int64

**Filter ind dataset for people with income more than 50K**

`here I want to know the working hours of these people `

In [ ]:
ind_50 = ind[ind['label'] == ' >50K' ]
ind_50.head()

index_adult  age     education       marital_status       occupation  \
4              11   30     Bachelors   Married-civ-spouse   Prof-specialty   
2467         8124   36   Prof-school        Never-married   Prof-specialty   
3063         9939   43   Prof-school   Married-civ-spouse   Prof-specialty   
3265        10590   35   Prof-school        Never-married   Prof-specialty   
3285        10661   59   Prof-school   Married-civ-spouse   Prof-specialty   

                     race    sex native_country  label  
4      Asian-Pac-Islander   Male          India   >50K  
2467                Other   Male          India   >50K  
3063   Asian-Pac-Islander   Male          India   >50K  
3265   Asian-Pac-Islander   Male          India   >50K  
3285   Asian-Pac-Islander   Male          India   >50K

* indices are intact with their rows, just like address

**Very Impressive, these people are earning really good, Lets try to know how much they work, since we dont have 'hours_per_week' in this data, we will bring it from df with the help of indices**

In [ ]:
adult['hours_per_week'][ind_50.index_adult]

11       40
8124     40
9939     50
10590    40
10661    40
13551    40
16923    60
17834    50
21128    50
24154    50
25739    40
26356    40
28264    70
28433    55
28452    40
28798    50
31327    60
Name: hours_per_week, dtype: int64

* Indices made it very easy to bring more information easily, so we can read it as filtering adult['hours_per_week'] on the address index_default of ind_50, its the same index

**Lets try to find the mean and compare it with the other people too **

In [ ]:
adult['hours_per_week'][ind_50.index_adult].mean()

47.94117647058823

* Just with the use of index_adult we were able to bring an another column information easily
* Index make filtering very easy and also give you space to move forward and backward in your data 

## one last use of index for this intro exercise

**Filtering complimentry set from the data, just like train and test from total datset ** <br>
**we are slicing that part of ind, which is not in ind_50, i.e. people who are earning less than 50k**

In [ ]:
ind_05 = ind[~ind.index.isin(ind_50.index)]
ind_05.head()

index_adult  age   education          marital_status        occupation  \
312          1029   48     Masters   Married-spouse-absent             Sales   
637          2130   28        11th               Separated      Adm-clerical   
902          2989   44     Masters      Married-civ-spouse   Exec-managerial   
1365         4480   30   Bachelors      Married-civ-spouse             Sales   
1379         4523   57   Bachelors      Married-civ-spouse    Prof-specialty   

                     race      sex native_country   label  
312    Asian-Pac-Islander     Male          India   <=50K  
637    Asian-Pac-Islander   Female          India   <=50K  
902    Asian-Pac-Islander     Male          India   <=50K  
1365   Asian-Pac-Islander     Male          India   <=50K  
1379   Asian-Pac-Islander     Male          India   <=50K

In [ ]:
adult['hours_per_week'][ind_05.index_adult].mean()

38.7

> Nice ! It looks like people who earn 50k&more work more hours per week in this data sample